core/src/main/java/org/infinispan/statetransfer/StateTransferInterceptor.java

```diff
@@ -231,6 +225,16 @@ private Object handleTxCommand(TxInvocationContext ctx, TransactionBoundaryComma
          }
       } else {
          if (currentTopologyId() > command.getTopologyId()) {
+            //in some cases, the node will not retry the transaction in this node and then it leaks.
+            //if it is safe, we rollback the transaction. If the coordinator retries on this node, no side effect is created.
```

Comment:
Hmmm, I missed that you also check the tx having acquired any locks. The comment about when the local rollback is needed is really vague, though, can you expand a bit on which scenarios do require a local rollback?